### PR TITLE
MM-14289 & MM-14884 Push notification acknowledge id and include sender name

### PR DESCRIPTION
**Pull request by enahum**
#### Summary

This PR adds support for receipt delivery by including an acknowledge identifier attached to every notification, also includes a new Database table to keep track of the notification status specifying if the notification was sent or not, the reason and timestamps as well as when it was received on the device.

Included in this PR a property in the notification that specifies the sender name.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-14289
https://mattermost.atlassian.net/browse/MM-14884

### CODE REVIEW

File: model/push_notification.go [hunk 1]
```diff
@@ -24,9 +25,32 @@ const (
 	CATEGORY_CAN_REPLY = "CAN_REPLY"
 
 	MHPNS = "https://push.mattermost.com"
+
+	PUSH_SEND_SUCCESS = "Successful"
+	PUSH_SEND_ERROR   = "Error:"
```
Comment (jespino) [inline model/push_notification.go]: nit: Maybe is more consistent remove the `:` from the string here, and add in the concatenation where it is used.

Comment (enahum) [reply model/push_notification.go]: done

File: api4/system.go [hunk 2]
```diff
@@ -317,3 +319,28 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(model.MapToJson(m)))
 	return
 }
+
+func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
+	ack := model.PushNotificationAckFromJson(r.Body)
+
+	if !*c.App.Config().EmailSettings.SendPushNotifications {
+		c.Err = model.NewAppError("pushNotificationAck", "api.push_notification.disabled.app_error", nil, "", http.StatusForbidden)
```
Comment (hmhealey) [inline api4/system.go]: We usually use a 501 for routes that are disabled by the settings
```suggestion
		c.Err = model.NewAppError("pushNotificationAck", "api.push_notification.disabled.app_error", nil, "", http.StatusNotImplemented)
```

File: app/notification.go [hunk 3]
```diff
@@ -306,6 +306,18 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 					(channelNotification || hereNotification || allNotification),
 					replyToThreadType,
 				)
+			} else {
+				// register that a notification was not sent
+				notificationRegistry := model.NotificationRegistry{
```
Comment (hmhealey) [inline app/notification.go]: Why do we want to track these? It just means that the user was online or something that prevented the notification

File: app/notification.go [hunk 4]
```diff
@@ -691,7 +715,7 @@ type postNotification struct {
 func (n *postNotification) GetChannelName(userNameFormat string, excludeId string) string {
 	switch n.channel.Type {
 	case model.CHANNEL_DIRECT:
-		return fmt.Sprintf("@%s", n.sender.GetDisplayName(userNameFormat))
+		return n.sender.GetDisplayName(userNameFormat)
```
Comment (hmhealey) [inline app/notification.go]: Should we leave this as it is for supporting older mobile apps? As it is, I assume the channel name won't be displayed on updated apps.

File: app/notification.go (hunk 3 above)
Comment (enahum) [reply app/notification.go]: Yes we want to track this when the user is online or the preferences of the user make it so the user don't receive the notification, this is in order to have a way to identify why the user did not receive the notification.

File: app/notification.go (hunk 4 above)
Comment (enahum) [reply app/notification.go]: Nope, this was changed in #10298 but they missed this one

**Pull request merged**